services/worker-manager/src/providers/azure.js

```diff
@@ -0,0 +1,447 @@
+    // extra information.
+    const error = () => new ApiError('Signature validation error');
+
+    // workerIdentityProof is a signed message
+
+    // We need to check that:
+    // 1. The embedded document was signed with the private key corresponding to the
+    //    embedded public key
+    // 2. The embedded public key has a proper certificate chain back to a trusted CA
+    // 3. The embedded message contains the vmId that matches the worker making the
+    //    worker making the request
+
+    // signature is base64-encoded DER-format PKCS#7 / CMS message
+
+    // decode base64, load DER, extract PKCS#7 message
+    let decodedMessage = Buffer.from(document, 'base64');
+    let message;
+    try {
+      let asn1 = forge.asn1.fromDer(forge.util.createBuffer(decodedMessage));
+      message = forge.pkcs7.messageFromAsn1(asn1);
+    } catch (err) {
+      this.monitor.warning('Error extracting PKCS#7 message', {error: err.toString()});
+      throw error();
+    }
+
+    let content, crt, pem, sig;
+    // get message content, signing certificate, and signature
+    try {
+      // in testing, message.content is empty, so we access the raw ASN1 structure
+      content = message.rawCapture.content.value[0].value;
+      // convert to pem for convenience
+      // there may be more than one certificate, we only look at the first one
+      crt = message.certificates[0];
+      pem = forge.pki.publicKeyToPem(crt.publicKey);
+      sig = message.rawCapture.signature;
+    } catch (err) {
+      this.monitor.warning('Error extracting PKCS#7 message content', {error: err.toString()});
+      throw error();
+    }
+
+    // verify that the message is properly signed
+    try {
+      let verifier = crypto.createVerify('RSA-SHA256');
+      verifier.update(Buffer.from(content));
+      assert(verifier.verify(pem, sig, 'binary'), true);
+    } catch (err) {
+      this.monitor.warning('Error verifying PKCS#7 message signature');
+      throw error();
+    }
+
+    // verify that the embedded certificate has proper chain of trust
+    try {
+      assert(forge.pki.verifyCertificateChain(this.caStore, [crt]), true);
+    } catch (err) {
+      this.monitor.warning('Error verifying certificate chain', {error: err.toString()});
+      throw error();
+    }
+
+    // verify that the embedded vmId matches what the worker is sending
+    try {
+      let {vmId} = JSON.parse(content);
+      assert(vmId === worker.workerId);
+    } catch (err) {
+      this.monitor.warning('Encountered vmId mismatch', {error: err.toString()});
+      throw error();
+    }
+
+    this.monitor.log.workerRunning({
+      workerPoolId: workerPool.workerPoolId,
+      providerId: this.providerId,
+      workerId: worker.workerId,
+    });
+    await worker.modify(w => {
+      w.lastModified = new Date();
+      w.state = this.Worker.states.RUNNING;
+    });
+
+    // assume for the moment that workers self-terminate before 96 hours
```

Comment:
What does this assumption imply exactly? If the instance doesn't self-terminate within 96 hours, what does that mean?

Reply:
We're giving out temporary credentials, so the credentials will expire and be unusable after 96 hours. This makes sense for spot instances in GCP and AWS because they are expected to be short lived... but I've realized that Azure does not guarantee any kind of shutdown timing, so this could be an issue.

Reply:
It isn't a security issue however, just a usability one. The creds do expire and the worker is just useless at that point

Reply:
Ok totally, sounds good.